File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/Party.java
// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.daml.ledger.javaapi.data;

import com.daml.ledger.api.v1.ValueOuterClass;
import java.util.Objects;
import org.checkerframework.checker.nullness.qual.NonNull;

public final class Party extends Value {

  private final String value;

  public Party(@NonNull String value) {
    this.value = value;
  }

  @Override
  public ValueOuterClass.Value toProto() {
    return ValueOuterClass.Value.newBuilder().setParty(this.value).build();
  }

  @NonNull
  public String getValue() {
    return value;
  }

  @Override
  public String toString() {
    return "Party{" + "value='" + value + '\'' + '}';
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Party party = (Party) o;
    return Objects.equals(value, party.value);
  }

  @Override
  public int hashCode() {

    return Objects.hash(value);
  }
}

File: sdk/canton/community/ledger-test-tool/suites/lf-v2.1/src/main/scala/com/daml/ledger/api/testtool/infrastructure/Party.scala
// Copyright (c) 2026 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.daml.ledger.api.testtool.infrastructure

import com.daml.ledger.api.v2.crypto as lapicrypto
import com.daml.ledger.api.v2.crypto.SignatureFormat.SIGNATURE_FORMAT_RAW
import com.daml.ledger.javaapi.data.Party as ApiParty
import com.digitalasset.canton.config.RequireTypes.PositiveInt
import com.digitalasset.canton.crypto.Fingerprint
import com.google.protobuf.ByteString

import java.security.{KeyPair, Signature}
import scala.language.implicitConversions

sealed trait Party {
  def underlying: ApiParty
  def initialSynchronizers: List[String]
}
final case class LocalParty(underlying: ApiParty, initialSynchronizers: List[String]) extends Party
final case class ExternalParty(
    underlying: ApiParty,
    initialSynchronizers: List[String],
    signingFingerprint: Fingerprint,
    signingKeyPair: KeyPair,
    signingThreshold: PositiveInt,
) extends Party {
  def sign(data: ByteString): ByteString = {
    val signatureInstance = Signature.getInstance("Ed25519")
    signatureInstance.initSign(signingKeyPair.getPrivate)
    signatureInstance.update(data.toByteArray)
    ByteString.copyFrom(signatureInstance.sign())
  }

  def signProto(data: ByteString): lapicrypto.Signature =
    lapicrypto.Signature(
      format = SIGNATURE_FORMAT_RAW,
      signature = sign(data),
      signedBy = signingFingerprint.toProtoPrimitive,
      signingAlgorithmSpec = lapicrypto.SigningAlgorithmSpec.SIGNING_ALGORITHM_SPEC_ED25519,
    )
}

object Party {

  def external(
      value: String,
      signingFingerprint: Fingerprint,
      signingKeyPair: KeyPair,
      signingThreshold: PositiveInt,
      initialSynchronizers: List[String] = List.empty,
  ): ExternalParty =
    ExternalParty(
      new ApiParty(value),
      initialSynchronizers,
      signingFingerprint,
      signingKeyPair,
      signingThreshold,
    )

  def apply(value: String, initialSynchronizers: List[String] = List.empty): Party =
    LocalParty(new ApiParty(value), initialSynchronizers)
  implicit def toApiParty(party: Party): ApiParty = party.underlying
  implicit def toApiString(party: Party): String = party.underlying.getValue
}
